fdb-extensions/src/main/java/com/apple/foundationdb/async/hnsw/HNSW.java

```diff
@@ -0,0 +1,1945 @@
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static com.apple.foundationdb.async.MoreAsyncUtil.forEach;
+import static com.apple.foundationdb.async.MoreAsyncUtil.forLoop;
+
+/**
+ * An implementation of the Hierarchical Navigable Small World (HNSW) algorithm for
+ * efficient approximate nearest neighbor (ANN) search.
+ * <p>
+ * HNSW constructs a multi-layer graph, where each layer is a subset of the one below it.
+ * The top layers serve as fast entry points to navigate the graph, while the bottom layer
+ * contains all the data points. This structure allows for logarithmic-time complexity
+ * for search operations, making it suitable for large-scale, high-dimensional datasets.
+ * <p>
+ * This class provides methods for building the graph ({@link #insert(Transaction, Tuple, RealVector)})
+ * and performing k-NN searches ({@link #kNearestNeighborsSearch(ReadTransaction, int, int, RealVector)}).
+ * It is designed to be used with a transactional storage backend, managed via a {@link Subspace}.
+ *
+ * @see <a href="https://arxiv.org/abs/1603.09320">Efficient and robust approximate nearest neighbor search using Hierarchical Navigable Small World graphs</a>
+ */
+@API(API.Status.EXPERIMENTAL)
+@SuppressWarnings("checkstyle:AbbreviationAsWordInName")
+public class HNSW {
+    @Nonnull
+    private static final Logger logger = LoggerFactory.getLogger(HNSW.class);
+
+    public static final int MAX_CONCURRENT_NODE_READS = 16;
+    public static final int MAX_CONCURRENT_NEIGHBOR_FETCHES = 3;
+    public static final int MAX_CONCURRENT_SEARCHES = 10;
+    public static final long DEFAULT_RANDOM_SEED = 0L;
+    @Nonnull public static final Metric DEFAULT_METRIC = Metric.EUCLIDEAN_METRIC;
+    public static final boolean DEFAULT_USE_INLINING = false;
+    public static final int DEFAULT_M = 16;
+    public static final int DEFAULT_M_MAX = DEFAULT_M;
+    public static final int DEFAULT_M_MAX_0 = 2 * DEFAULT_M;
+    public static final int DEFAULT_EF_CONSTRUCTION = 200;
+    public static final boolean DEFAULT_EXTEND_CANDIDATES = false;
+    public static final boolean DEFAULT_KEEP_PRUNED_CONNECTIONS = false;
+
+    // RaBitQ
+    public static final boolean DEFAULT_USE_RABITQ = false;
+    public static final int DEFAULT_RABITQ_NUM_EX_BITS = 4;
+
+    @Nonnull
+    public static final ConfigBuilder DEFAULT_CONFIG_BUILDER = new ConfigBuilder();
+
+    @Nonnull
+    private final Random random;
+    @Nonnull
+    private final Subspace subspace;
+    @Nonnull
+    private final Executor executor;
+    @Nonnull
+    private final Config config;
+    @Nonnull
+    private final OnWriteListener onWriteListener;
+    @Nonnull
+    private final OnReadListener onReadListener;
+
+    /**
+     * Configuration settings for a {@link HNSW}.
+     */
+    @SuppressWarnings("checkstyle:MemberName")
+    public static class Config {
```

Comment:
This class is pretty big, and teamscale is suggesting moving the config out anyway. It might be a good idea just to keep things contained